[Debug Mode] Add transaction and violation data to debug mode

pac-guerreiro opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$https://github.com/Expensify/App/issues/50335
PROPOSAL:
- Add transaction debug page, with details, json and violations tabs
- Add transaction violation page
- Add button to redirect user from report debug page to transaction report page
- Add button to redirect user from report action debug page to transaction report page

### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

1. Enable debug mode in Troubleshoot
2. Open an expense thread (create one if you don't have any)
3. Enter debug view then check if there's a button named `View transaction` in details tab
4. Click the button and check the `details`, `json` and `violations` tabs
5. Confirm that any field you modify in `details` tab is saved correctly
6. On `violations` tab, try creating and modifying a violation and check that the UI shows your changes in real time

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

Same as tests

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
7. Upload an image via copy paste
8. Verify a modal appears displaying a preview of that image
--->

Same as tests

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->
https://github.com/user-attachments/assets/3b130905-e03f-4dad-b835-38d7d2c2ffd9

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->
https://github.com/user-attachments/assets/30b6b982-74df-45fe-9b3f-fa36220444e3

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->
https://github.com/user-attachments/assets/f2f017e6-cd7d-46f9-83f9-ab36ef9429db

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->
https://github.com/user-attachments/assets/b92f2835-c36e-4cb7-8783-0d67669bba93

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->
https://github.com/user-attachments/assets/2a0bf059-4056-4bb4-9883-5c4ee6f4d372

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->
https://github.com/user-attachments/assets/841f4c85-b612-451e-90f2-6b7af2c474b8

</details>

pac-guerreiro committed:
feat(debug mode): add debug transaction page

pac-guerreiro commented:
Todo:
- [ ] Create `DebugUtils.validateTransactionDraftProperty` plus unit tests
- [ ] Finish transaction violations tab
- [ ] Add screen recordings
- [ ] Add QA testing steps

pac-guerreiro committed:
chore: create validateTransactionDraftProperty

pac-guerreiro commented:
Today's update:
- [x] Create `DebugUtils.validateTransactionDraftProperty`

Todo:
- [ ] Add remaining property validations to `DebugUtils.validateTransactionDraftProperty`
- [ ] Add unit tests for `DebugUtils.validateTransactionDraftProperty`
- [ ] Do some manual tests by modifying transaction data and saving it
- [ ] Finish transaction violations tab
- [ ] Create button in debug report page to redirect to debug transaction page
- [ ] Add screen recordings
- [ ] Add QA testing steps

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/50335-add-transaction-and-violation-data-to-debug-mode

# Conflicts:
#	src/libs/DebugUtils.ts

pac-guerreiro committed:
chore(debug mode): finish DebugUtils.validateTransactionDraftData and add unit tests

pac-guerreiro commented:
Today's update:
- [x] Add remaining property validations to `DebugUtils.validateTransactionDraftProperty`
- [x] Add unit tests for `DebugUtils.validateTransactionDraftProperty`
- [x] Do some manual tests by modifying transaction data and saving it

Todo:
- [ ] Create `TRANSACTION_FORM_INPUT_IDS`
- [ ] Finish transaction violations tab
- [ ] Create button in debug report page to redirect to debug transaction page
- [ ] Add screen recordings
- [ ] Add QA testing steps

pac-guerreiro committed:
chore(debug mode): add debug transaction types

pac-guerreiro committed:
feat(debug mode): add links for transaction and report

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/50335-add-transaction-and-violation-data-to-debug-mode

pac-guerreiro commented:
Today's update:
- [x] Create `TRANSACTION_FORM_INPUT_IDS`
- [x] Create button in debug report page to redirect to debug transaction page
- [x] Create button in debug report action page to redirect to debug transaction page
- [x] Create button in debug transaction to redirect to report

Todo:
- [ ] Finish transaction violations tab
- [ ] Add screen recordings
- [ ] Add QA testing steps

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/50335-add-transaction-and-violation-data-to-debug-mode

pac-guerreiro committed:
chore(debug mode): add transaction violation creation page

pac-guerreiro commented:
Today's update:
- [x] Add transaction violation creation page

Todo:
- [ ] Fix issue where newly created transaction violation replaces all existing transaction violations
- [ ] Add transaction violation page to allow users to modify data, view JSON data representation and delete data
- [ ] Finish transaction violations tab
- [ ] Add screen recordings
- [ ] Add QA testing steps

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/50335-add-transaction-and-violation-data-to-debug-mode

pac-guerreiro committed:
chore(debug mode): add debug transaction violation page and finish debug transaction violations tab

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/50335-add-transaction-and-violation-data-to-debug-mode

# Conflicts:
#	src/libs/DebugUtils.ts

pac-guerreiro commented:
Today's update:
- [x] Fix issue where newly created transaction violation replaces all existing transaction violations
- [x] Add transaction violation page to allow users to modify data, view JSON data representation and delete data
- [x] Finish transaction violations tab

Todo:
- [ ] Fix stale transaction violation data after updating a violation until the app is refreshed
- [ ] Add screen recordings
- [ ] Add QA testing steps

pac-guerreiro committed:
chore(debug mode): add debug transaction violation draft form types

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/50335-add-transaction-and-violation-data-to-debug-mode

pac-guerreiro committed:
chore(debug mode): fix mutating Onyx data causing set to fail and add transaction violations form types

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/50335-add-transaction-and-violation-data-to-debug-mode

pac-guerreiro commented:
Today's update:
- [x] Fix stale transaction violation data after updating a violation until the app is refreshed
- [x] Refactor debug details constant field options. There are field name collisions between data types

Todo:
- [ ] Add screen recordings
- [ ] Add QA testing steps

pac-guerreiro committed:
refactor(debug mode): debug details constant fields

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/50335-add-transaction-and-violation-data-to-debug-mode

pac-guerreiro commented:
@DylanDylann @puneetlath 

The PR is now ready for review 😄

DylanDylann commented:
@pac-guerreiro Type errors please

pac-guerreiro committed:
refactor: make code more readable

pac-guerreiro committed:
fix: virtualized list error

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/50335-add-transaction-and-violation-data-to-debug-mode

fabioh8010 commented on src/pages/Debug/ConstantSelector.tsx:
```diff
@@ -21,9 +22,11 @@ type ConstantSelectorProps = {
     // eslint-disable-next-line react/no-unused-prop-types
     inputID: string;
+
+    formType: ValueOf<typeof CONST.DEBUG.FORMS>;
```
Please add prop description

fabioh8010 commented on src/pages/Debug/DebugDetails.tsx:
```diff
@@ -16,17 +17,19 @@ import type {ObjectType, OnyxDataType} from '@libs/DebugUtils';
 import DateTimeSelector from './DateTimeSelector';
 
 type DebugDetailsProps = {
+    formType: ValueOf<typeof CONST.DEBUG.FORMS>;
```
Please add prop description

fabioh8010 commented on src/pages/Debug/Transaction/DebugTransactionViolations.tsx:
```diff
@@ -0,0 +1,57 @@
+};
+
+function DebugTransactionViolations({transactionID}: DebugTransactionViolationsProps) {
+    const [transactionViolations] = useOnyx(`${ONYXKEYS.COLLECTION.TRANSACTION_VIOLATIONS}${transactionID}`, {});
```
```suggestion
    const [transactionViolations] = useOnyx(`${ONYXKEYS.COLLECTION.TRANSACTION_VIOLATIONS}${transactionID}`);
```

fabioh8010 commented on src/pages/Debug/TransactionViolation/DebugTransactionViolationCreatePage.tsx:
```diff
@@ -0,0 +1,130 @@
+                                numberOfLines={18}
+                                multiline
+                                value={draftTransactionViolation}
+                                onChangeText={(updatedJSON) => {
+                                    try {
+                                        DebugUtils.validateTransactionViolationJSON(updatedJSON);
+                                        setError('');
+                                    } catch (e) {
+                                        const {cause, message} = e as SyntaxError;
+                                        setError(cause ? translate(message as TranslationPaths, cause as never) : message);
+                                    } finally {
+                                        setDraftTransactionViolation(updatedJSON);
+                                    }
+                                }}
```
Could you extract this into separate function?

fabioh8010 commented on src/pages/Debug/TransactionViolation/DebugTransactionViolationCreatePage.tsx:
```diff
@@ -0,0 +1,130 @@
+                            success
+                            text={translate('common.save')}
+                            isDisabled={!draftTransactionViolation || !!error}
+                            onPress={() => {
+                                const parsedTransactionViolation = DebugUtils.stringToOnyxData(draftTransactionViolation, 'object') as TransactionViolation;
+                                Debug.mergeDebugData(`${ONYXKEYS.COLLECTION.TRANSACTION_VIOLATIONS}${transactionID}`, [...(transactionViolations ?? []), parsedTransactionViolation]);
+                                Navigation.navigate(ROUTES.DEBUG_TRANSACTION_TAB_VIOLATIONS.getRoute(transactionID));
+                            }}
```
Same

fabioh8010 commented on src/pages/Debug/TransactionViolation/DebugTransactionViolationPage.tsx:
```diff
@@ -0,0 +1,86 @@
+                                <DebugDetails
+                                    formType={CONST.DEBUG.FORMS.TRANSACTION_VIOLATION}
+                                    data={transactionViolation}
+                                    onSave={(data) => {
+                                        const updatedTransactionViolations = [...(transactionViolations ?? [])];
+                                        updatedTransactionViolations.splice(Number(index), 1, data as TransactionViolation);
+                                        Debug.setDebugData(`${ONYXKEYS.COLLECTION.TRANSACTION_VIOLATIONS}${transactionID}`, updatedTransactionViolations);
+                                    }}
+                                    onDelete={() => {
+                                        const updatedTransactionViolations = [...(transactionViolations ?? [])];
+                                        updatedTransactionViolations.splice(Number(index), 1);
+                                        Debug.setDebugData(`${ONYXKEYS.COLLECTION.TRANSACTION_VIOLATIONS}${transactionID}`, updatedTransactionViolations);
+                                    }}
```
Same

fabioh8010 commented on src/libs/DebugUtils.ts:
```diff
@@ -271,7 +294,7 @@ function validateBoolean(value: string) {
  * Validates if a string is a valid representation of a date.
  */
 function validateDate(value: string) {
```
Did you update the unit tests to cover the changes?

pac-guerreiro replied:
I did not but I'll confirm if it's needed, thanks! 😄

fabioh8010 commented on src/libs/DebugUtils.ts:
```diff
@@ -355,8 +378,8 @@ function validateArray(
  */
-function validateObject(value: string, type: ObjectType, collectionIndexType?: 'string' | 'number') {
-    if (value === 'undefined') {
+function validateObject<T extends Record<string, unknown>>(value: string, type: ObjectType<T>, collectionIndexType?: 'string' | 'number') {
```
Did you update the unit tests to cover the changes?

pac-guerreiro replied:
I did not but I'll confirm if it's needed, thanks! 😄

pac-guerreiro commented:
@DylanDylann typecheck errors were resolved 😄

pac-guerreiro committed:
refactor: apply suggestions

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/50335-add-transaction-and-violation-data-to-debug-mode

# Conflicts:
#	tests/utils/collections/transaction.ts

pac-guerreiro commented:
@fabioh8010 all feedback addressed, thanks! 😄 

Still waiting for @DylanDylann review 😄

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/50335-add-transaction-and-violation-data-to-debug-mode

pac-guerreiro committed:
fix: jest unit tests failing

pac-guerreiro commented:
@puneetlath 

I fixed the jest unit tests that were failing

Still waiting on additional review from @DylanDylann 😄

pac-guerreiro committed:
chore: add missing test cases for debug utils

pac-guerreiro committed:
fix(debug mode): view transaction now showing for some expense chats

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/50335-add-transaction-and-violation-data-to-debug-mode

pac-guerreiro commented:
@puneetlath @DylanDylann @fabioh8010 

Today's update:

- [x] Added missing test cases for `validateConstantEnum` and `validateObject` (thanks for pointing this @fabioh8010 )
- [x] Fixed some expense chats not showing `View Transaction` button in debug panel

fabioh8010 commented on src/pages/Debug/TransactionViolation/DebugTransactionViolationCreatePage.tsx:
```diff
@@ -66,6 +66,27 @@ function DebugTransactionViolationCreatePage({
     const [draftTransactionViolation, setDraftTransactionViolation] = useState<string>(getInitialTransactionViolation());
     const [error, setError] = useState<string>();
 
+    const onEditJSON = useCallback(
```
Avoid naming with `on` prefixes for event handles, https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#event-handlers

pac-guerreiro replied:
Thanks for the tip @fabioh8010  😄

fabioh8010 commented on src/pages/Debug/TransactionViolation/DebugTransactionViolationCreatePage.tsx:
```diff
@@ -66,6 +66,27 @@ function DebugTransactionViolationCreatePage({
+        [translate],
+    );
+
+    const onSave = useCallback(() => {
```
Same

fabioh8010 commented on src/pages/Debug/TransactionViolation/DebugTransactionViolationPage.tsx:
```diff
@@ -33,6 +33,21 @@ function DebugTransactionViolationPage({
     const transactionViolation = useMemo(() => transactionViolations?.[Number(index)], [index, transactionViolations]);
     const styles = useThemeStyles();
 
+    const onSave = useCallback(
```
Same

fabioh8010 commented on src/pages/Debug/TransactionViolation/DebugTransactionViolationPage.tsx:
```diff
@@ -33,6 +33,21 @@ function DebugTransactionViolationPage({
+        [index, transactionID, transactionViolations],
+    );
+
+    const onDelete = useCallback(() => {
```
Same

fabioh8010 reviewed (commented):
minor comments

pac-guerreiro committed:
refactor: apply suggestions

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/50335-add-transaction-and-violation-data-to-debug-mode

pac-guerreiro commented:
@fabioh8010 feedback applied, thanks 😄

DylanDylann commented:
BUG: Don't add a default value in the comment field. Let's add {} as default value

https://github.com/user-attachments/assets/88bf034f-f469-4bc6-8a8a-4591917cbb1a

DylanDylann commented:
The same bug on the report debug page, the validation doesn't match the Onyx data

https://github.com/user-attachments/assets/5931ac76-98b1-48b0-b594-583d8325f72b

DylanDylann commented:
https://github.com/user-attachments/assets/b37d818d-255a-4a6b-b2b4-082947ea728d

Category and Tag field should be a selected picker instead of a text

DylanDylann commented on src/pages/Debug/TransactionViolation/DebugTransactionViolationPage.tsx:
```diff
@@ -0,0 +1,93 @@
+    const deleteTransactionViolation = useCallback(() => {
+        const updatedTransactionViolations = [...(transactionViolations ?? [])];
+        updatedTransactionViolations.splice(Number(index), 1);
+        Debug.setDebugData(`${ONYXKEYS.COLLECTION.TRANSACTION_VIOLATIONS}${transactionID}`, updatedTransactionViolations);
```
same above

pac-guerreiro replied:
Initially I tried using that function but it wasn't saving the data properly, but not that I tried it again it seems to be working!

I think it's working now because I ended up doing some changes at the time, so I'm removing `Debug.setDebugData`

DylanDylann commented on src/pages/Debug/TransactionViolation/DebugTransactionViolationPage.tsx:
```diff
@@ -0,0 +1,93 @@
+        (data: Record<string, unknown>) => {
+            const updatedTransactionViolations = [...(transactionViolations ?? [])];
+            updatedTransactionViolations.splice(Number(index), 1, data as TransactionViolation);
+            Debug.setDebugData(`${ONYXKEYS.COLLECTION.TRANSACTION_VIOLATIONS}${transactionID}`, updatedTransactionViolations);
```
could we use Debug.mergeDebugData?

pac-guerreiro replied:
Same as above

DylanDylann commented:
The validation doesn't match the Onyx data

https://github.com/user-attachments/assets/1606deb4-b792-41f0-b7c5-0718c6f5ddf6

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/50335-add-transaction-and-violation-data-to-debug-mode

pac-guerreiro commented:
> Screen.Recording.2024-10-30.at.10.53.27.mov 
> Category and Tag field should be a selected picker instead of a text

Implementing the picker for Category should be doable but I'll need to investigate more! For Tag field it can be complicated since there are dependant tags with at least 3 levels 😅

pac-guerreiro commented:
> On transaction validation eidt -> data field. The validation doesn't match the Onyx data.
> 
>  Screen.Recording.2024-10-30.at.11.38.00.mov

@DylanDylann can you share the data json with me? I think there might be a mismatch between our typescript types and what's coming from the API

pac-guerreiro committed:
refactor(debug mode): apply suggestions

pac-guerreiro committed:
refactor: replace union with value of constant

pac-guerreiro committed:
refactor(debug mode): enforce property validation based on onyx data type

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/50335-add-transaction-and-violation-data-to-debug-mode

pac-guerreiro commented:
@DylanDylann I addressed most of the feedback you left today! There's just a couple of questions that I'll need answers from you 😄 

Today I did a major refactor on DebugUtils, in order to help prevent property validations from failing in the future.

pac-guerreiro commented:
@DylanDylann It seems that as a bonus I broke some tests and types 😅 I'll address these tomorrow!

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/50335-add-transaction-and-violation-data-to-debug-mode

pac-guerreiro committed:
chore: resolve eslint and jest issues

pac-guerreiro committed:
fix: add missing property to transaction violation data type

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/50335-add-transaction-and-violation-data-to-debug-mode

pac-guerreiro committed:
chore: resolve eslint issues

pac-guerreiro commented:
Today's update:

- [x] Fixed eslint issues
- [x] Fixed transaction violation validation issue reported by @DylanDylann
- [x] Fixed unit tests failing

Todo:

- [] Add selector for transaction category
- [] Investigate how much work it takes to make a selector for transaction tags

DylanDylann commented:
>  Investigate how much work it takes to make a selector for transaction tags

If this is complicated, I think we can skip

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/50335-add-transaction-and-violation-data-to-debug-mode

pac-guerreiro committed:
feat(debug mode): add ability to select category and tag from a picker

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/50335-add-transaction-and-violation-data-to-debug-mode

pac-guerreiro commented:
Sorry for the late update, here's what I did yesterday:

- [x] Add selector for transaction category
- [x] Investigate how much work it takes to make a selector for transaction tags 
- [x] Add selector for transaction tag with only one level

Todo:

- [ ] Modify transaction tag selector to allow editing multi-level tags

pac-guerreiro commented:
Here is a preview of the category and tag pickers in action:

https://github.com/user-attachments/assets/154ea6d2-34e8-4e0b-89f4-684cfcc2cc83

pac-guerreiro committed:
feat(debug mode): add ability to pick multi-level transaction tags

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/50335-add-transaction-and-violation-data-to-debug-mode

# Conflicts:
#	src/languages/en.ts
#	src/languages/es.ts
#	src/pages/Debug/DebugDetails.tsx
#	src/pages/Debug/Report/DebugReportPage.tsx

pac-guerreiro commented:
@DylanDylann I added the ability to edit multi-level tags in the picker 😄 

Let me know if you find any problems!

DylanDylann commented:
Back to this PR, today

DylanDylann commented:
@pac-guerreiro Can't select VND currency on transaction debug mode

<img width="1512" alt="Screenshot 2024-11-06 at 16 19 54" src="https://github.com/user-attachments/assets/f831bc18-03ef-4ade-91ba-b67f81979690">

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/50335-add-transaction-and-violation-data-to-debug-mode

pac-guerreiro committed:
fix(debug mode): currency selector has lesser options than expected

pac-guerreiro commented:
@DylanDylann I just pushed the fix for the currency picker not showing all available currencies!

There are some TODOs that I left in `DebugUtils.ts`, which I need to resolve before this gets merged

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/50335-add-transaction-and-violation-data-to-debug-mode

pac-guerreiro committed:
fix: type definitions for arrays and objects used in validations

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/50335-add-transaction-and-violation-data-to-debug-mode

pac-guerreiro committed:
fix: missing array validation for previousMessage

pac-guerreiro commented:
Today I resolved most of the typescript issues that were annotated with the TODO comments. There's still one issue that I need to resolve regarding the keys of a union type, which I'll try to fix tomorrow.

pac-guerreiro committed:
fix: remaining type definition issue

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/50335-add-transaction-and-violation-data-to-debug-mode

# Conflicts:
#	src/libs/DebugUtils.ts
#	src/types/onyx/Report.ts

pac-guerreiro committed:
fix: linter issues

pac-guerreiro commented:
Yesterday I resolved all the typescript issues present in the PR

DylanDylann commented:
@pac-guerreiro Many thanks. I will try to review again on this weekend

DylanDylann commented:
@pac-guerreiro On the tag picker, let's remove the save button. I think we should update the value and navigate go back right after the user click on the new value

https://github.com/user-attachments/assets/34448fff-5230-4a46-98d0-b9f649637da6

DylanDylann commented:
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/e3e409b6-e7ef-4977-a024-1a438e320a38



</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/958cb671-7496-4eba-bfb0-4e489b529254



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/c062f113-4fa2-48d4-8b63-d4a78a93b4c9



</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/a65e8cc1-6246-46db-ae60-a8ff11d5155f


<!-- add screenshots or videos here -->

</details>

DylanDylann commented:
@pac-guerreiro One left comment: https://github.com/Expensify/App/pull/50745#issuecomment-2467409302

fabioh8010 commented on src/libs/DebugUtils.ts:
```diff
@@ -419,88 +424,425 @@ function validateString(value: string) {
-    }
-    if (key === 'participantAccountIDs') {
-        return validateArray(value, 'number');
+    switch (key) {
```
Is possible to simplify this? For example, combining all string validations in one case:

```tsx
switch (key) {
        // all string cases
        case 'avatarUrl':
        case 'avatarFileName':
        case 'lastVisibleActionCreated':
        case 'lastReadCreated':
        case 'lastReadTime': {
            return validateString(value);
        }
```

fabioh8010 commented on src/libs/DebugUtils.ts:
```diff
@@ -510,49 +852,766 @@ function validateReportDraftProperty(key: keyof Report, value: string) {
-    }
-    if (key === 'message') {
-        return validateArray(value, {text: 'string', html: ['string', 'undefined'], type: 'string'});
+    switch (key) {
```
Same

fabioh8010 commented on src/libs/DebugUtils.ts:
```diff
@@ -510,49 +852,766 @@ function validateReportDraftProperty(key: keyof Report, value: string) {
     }
-    if (key === 'errors') {
-        return validateObject(value, {});
+    switch (key) {
```
Same

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/50335-add-transaction-and-violation-data-to-debug-mode

pac-guerreiro committed:
refactor: apply suggestions

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/50335-add-transaction-and-violation-data-to-debug-mode

pac-guerreiro commented:
@DylanDylann @fabioh8010 Thanks for your feedback!

I addressed all your suggestions, let me know if there's anything else that needs fixing/refactoring 😄

DylanDylann reviewed (approved):
Look good to me 💯

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/50335-add-transaction-and-violation-data-to-debug-mode

# Conflicts:
#	src/components/TagPicker/index.tsx
#	src/libs/DebugUtils.ts
#	src/pages/Debug/Report/DebugReportPage.tsx
#	src/types/onyx/Report.ts

pac-guerreiro commented:
@DylanDylann thanks! 🙏 I just resolved the conflicts 😄

puneetlath commented on src/types/onyx/Report.ts:
```diff
@@ -37,11 +37,14 @@ type PendingChatMember = {
 /** Report participant properties */
 type Participant = OnyxCommon.OnyxValueWithOfflineFeedback<{
+    /** Whether the participant is hidden */
+    hidden?: boolean;
```
Why are we adding all this to the report type? We recently worked in this issue to clean up the Report type to only store data that is sent by the server. https://github.com/Expensify/App/issues/51867

pac-guerreiro replied:
I added this because it's being returned from the server 😅 

<img width="328" alt="Screenshot 2024-11-14 at 18 16 19" src="https://github.com/user-attachments/assets/518288ef-8ff6-4900-81ff-3690c2fab3a0">

puneetlath replied:
Ah gotcha. Yeah "hidden" is being removed from the server as we speak. What about the other ones. Do we need those?

pac-guerreiro replied:
I just removed all the unused properties like agreed in https://github.com/Expensify/App/issues/51867 😄 

Thanks for the feedback @puneetlath 😄

fabioh8010 commented on src/components/TagPicker/index.tsx:
```diff
@@ -31,11 +31,13 @@ type TagPickerProps = {
     /** Should show the selected option that is disabled? */
     shouldShowDisabledAndSelectedOption?: boolean;
 
+    shouldOrderListByTagName?: boolean;
```
Please add prop description

fabioh8010 commented on src/pages/Debug/DebugTagPicker.tsx:
```diff
@@ -0,0 +1,77 @@
+import * as TransactionUtils from '@libs/TransactionUtils';
+import ONYXKEYS from '@src/ONYXKEYS';
+
+type DebugTagPickerProps = {
+    policyID: string;
+    tagName?: string;
+    onSubmit: (item: ListItem) => void;
+};
```
Please add descriptions for each prop

fabioh8010 commented on src/pages/Debug/Transaction/DebugTransactionViolations.tsx:
```diff
@@ -0,0 +1,57 @@
+import ROUTES from '@src/ROUTES';
+import type {TransactionViolation} from '@src/types/onyx';
+
+type DebugTransactionViolationsProps = {
+    transactionID: string;
+};
```
Please add descriptions for each prop

fabioh8010 commented on src/pages/Debug/ConstantPicker.tsx:
```diff
@@ -0,0 +1,62 @@
+import type {DebugForms} from './const';
+import {DETAILS_CONSTANT_FIELDS} from './const';
+
+type ConstantPickerProps = {
+    formType: string;
+    fieldName: string;
+    fieldValue?: string;
+    onSubmit: (item: ListItem) => void;
+};
```
Please add descriptions for each prop

puneetlath commented:
Looks like there are lint failures. Also, this PR is huge. I would love to see in the future, that if we find the PR is growing large like this, to separate it into multiple smaller PRs. One GH issue does not have to mean one PR. It's always a good think to break a big problem into smaller chunks. For example, I think this PR could have been broken down into smaller chunks by: 
- one PR for showing transaction debug page with transaction details
- one PR for showing violation debug page with violation details 
- one PR for the ability to select different violations
- one PR for the ability to create new violations

pac-guerreiro committed:
chore: apply linter

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/50335-add-transaction-and-violation-data-to-debug-mode

pac-guerreiro committed:
chore: remove unused properties

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/50335-add-transaction-and-violation-data-to-debug-mode

pac-guerreiro committed:
refactor: remove unused properties

pac-guerreiro committed:
chore: remove unused properties

pac-guerreiro committed:
chore: revert deletion of report property

pac-guerreiro committed:
chore: apply suggestions

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/50335-add-transaction-and-violation-data-to-debug-mode

# Conflicts:
#	src/libs/DebugUtils.ts
#	tests/utils/collections/reports.ts

pac-guerreiro commented:
@puneetlath 

Sorry for the delay, I just resolved the remaining conflicts 😄 

Today I noticed that debug report page was not saving data properly, can you confirm this?

pac-guerreiro committed:
fix: merge conflicts

puneetlath commented:
Thanks for resolving the conflicts! Looks like we have some typescript failures now.

puneetlath commented:
> Today I noticed that debug report page was not saving data properly, can you confirm this?

In this PR or in the live app?

pac-guerreiro committed:
fix: eslint issues

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/50335-add-transaction-and-violation-data-to-debug-mode

pac-guerreiro commented:
@puneetlath just resolved the typescript issues!

> In this PR or in the live app?
In main branch, so I guess at least in staging

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/50335-add-transaction-and-violation-data-to-debug-mode

pac-guerreiro committed:
fix(debug mode): problem updating nested objects in onyx

pac-guerreiro commented:
@puneetlath 

I just confirmed that the problem is also present on this PR. I just pushed a fix for it 😄 

The problem was that we we're saving debug data using `Onyx.merge`, which doesn't recognise implicit property removals from objects. So I replaced this method with `Onyx.set`, which replaces all the data.

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/50335-add-transaction-and-violation-data-to-debug-mode

# Conflicts:
#	src/libs/DebugUtils.ts

pac-guerreiro commented:
@puneetlath conflicts resolved 😄

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.
